### PR TITLE
[Feature] Implementation of noise protocol.

### LINKED ISSUE #48: Simulate readout errors
This allows to emulate the presence of readout errors in a shot-based simulation.

### PR DESCRIPTION
**Pull request by Roland-djee**
Closes #48, rework of #23.

@gvelikova This is ready for you to take over. As discussed, there are the following points:

- [x] Move `./errors/errors.py` -> `./exceptions/exceptions.py`
- [x] Change protocol occurrences to measurement from/in qadence.measurements to remove ambiguity including `DifferentiableBackend`
- [x] Harmonise across all backends
- [x] Test the sample corruption from backends and quantum model
- [x] Close #23 whenever you feel so
- [x] Test with other protocols that require `sample` ie. `Measurements`
- [x] Add on sampling with readout error in the documentation.

Hope that makes sense.

### CODE REVIEW

**Comment by gvelikova** [comment]
Docstrings and proper tests remaining too :).

**Comment by Roland-djee** [comment]
@gvelikova You can work on top of this if you've got time.

**Comment by Roland-djee** [comment]
Hey @gvelikova sorry it took me a bit of time to get that right. You can have a look at the issue we discussed.

**Comment by Roland-djee** [comment]
Hi @gvelikova sorry I didn't see that you've been working on this. I've pushed changes and tests. Please tell me what you think and we can agree on ho wto proceed further with your changes too.

File: qadence/errors/readout.py
Comment (gvelikova) [inline qadence/errors/readout.py:150]: As we discussed, this is the 'original' code used for the readout mitigation, but the other version should be able to corrupt bitstrings and return correct counts.

Comment (gvelikova) [reply qadence/errors/readout.py:150]: Tested out the commented function and it returns the same results.

**Comment by gvelikova** [comment]
@Roland-djee I uncommented the previous function and added very simplistic JS tests.

**Review by madagra — COMMENTED** [review]
Thanks, @gvelikova and @Roland-djee. It looks quite good to me. A couple of general comments:
* I am not a big fan of the name `Errors`, I would much rather have the usual `Noise` which is much more common and understandable in the community.
* I think the warning messages and compatibility of various options should be improved. For example, simulated noise now works for both `sample` and `expectation`, but only if measurement protocols are selected for the latter. This should be better communicated to the user and centralized in a single place, I think.
* Documentation is missing. Remember that one of the major points which decreased usage of shot-based measurements in the past was that there was practically no documentation. This really needs to be addressed.
* I see why an `ExperimentComposer` will be needed if we add also, e.g., error mitigation to the various configurations. It can just be a single separate option to the backends/models called, e.g., `Protocols` or `RealisticModel` where we centralize the measurements, errors and mitigation.

**Comment by madagra** [comment]
@gvelikova let's try to merge this one ASAP after adding the documentation.

File: qadence/backend.py
Comment (gvelikova) [reply qadence/backend.py:23]: This is a good point, maybe noise really would be better.

File: qadence/backends/braket/backend.py
Comment (Roland-djee) [reply qadence/backends/braket/backend.py:178]: Exactly. Thing is, I thought it would make sense to at least inform the user in case he/she falls here instead of carrying on and silently ignoring the error. As a matter of fact, it should also hold for other protocols IMHO. About resolving it, it is the sense of the experiment composer thing that we discussed.

File: qadence/backends/pulser/backend.py
Comment (Roland-djee) [reply qadence/backends/pulser/backend.py:37]: OK fair enough. This also relates to my comment above. Please ignore it in case.

Comment (Roland-djee) [reply qadence/backends/pulser/backend.py:37]: ~But: wouldn't this log the `backend.py` file similarly ?~ Ah nah,  it also logs the path.

File: qadence/backends/pyqtorch/backend.py
Comment (Roland-djee) [reply qadence/backends/pyqtorch/backend.py:190]: Well, the point being that if you use `error` + `measurement`, it proceeds through. If you use `errors` only it is ignored and the user is warned.

File: qadence/backends/pytorch_wrapper.py
Comment (Roland-djee) [reply qadence/backends/pytorch_wrapper.py:88]: Well, indeed. We are gradually moving towards a batch of protocols supported throughout Qadence. They should be unified and validated (as discussed).

File: qadence/errors/protocols.py
Comment (Roland-djee) [reply qadence/errors/protocols.py:12]: Not sure this is possible because we need the initialiser which I don't think we can get in a `StrEnum` ? Unless I'm wrong ofc.

File: qadence/errors/protocols.py
Comment (Roland-djee) [reply qadence/errors/protocols.py:19]: Indeed. That's not the case atm. I ended up in a situation where `shadow` was failing when no options were passed although I thought I checked the case. Will fix this.

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:36]: No we haven't done that yet. I guess I am assigned a task about performance in shadows. It can come together ?

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:51]: Isn't something like this in the qoolbox ?

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:52]: I assume this is provided by the `error_probability` arg @gvelikova ?

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:87]: I guess I wanted to make the experiment composer a pydantic model that captures protocol options as well.

File: qadence/ml_tools/models.py
Comment (Roland-djee) [reply qadence/ml_tools/models.py:217]: Good catch, thanks.

File: tests/qadence/test_error_models.py
Comment (Roland-djee) [reply tests/qadence/test_error_models.py:41]: Indeed. I need to expand a bit more on tests. I wanted to check first with `proba=1.0` to get a deterministic output for all bits to be flipped.

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:36]: @gvelikova

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:50]: @gvelikova

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:51]: @gvelikova

File: qadence/backend.py
Comment (madagra) [reply qadence/backend.py:23]: Well, actually yes. All errors come from noise in my point of view. Also readout which comes from noise in the measurement equipment.

File: qadence/backends/braket/backend.py
Comment (madagra) [reply qadence/backends/braket/backend.py:178]: Yes, makes sense.

File: qadence/errors/protocols.py
Comment (madagra) [reply qadence/errors/protocols.py:19]: Thanks, a separate MR can also be good. Let's not pollute this one too much.

File: qadence/errors/readout.py
Comment (madagra) [reply qadence/errors/readout.py:36]: Sure. Let's update the issue description.

File: qadence/errors/readout.py
Comment (madagra) [reply qadence/errors/readout.py:51]: Not really that general.

File: qadence/errors/readout.py
Comment (madagra) [reply qadence/errors/readout.py:87]: OK, makes sense. Can you create/update the issue for the experiment composer?

File: qadence/backends/pyqtorch/backend.py
Comment (madagra) [reply qadence/backends/pyqtorch/backend.py:190]: Yeah, this really needs to be better explained.

File: qadence/backend.py
Comment (madagra) [reply qadence/backend.py:29]: Not really, there are many modules with the same name.

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:52]: @gvelikova

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:63]: @gvelikova

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:114]: @gvelikova

File: tests/qadence/test_error_models.py
Comment (Roland-djee) [inline tests/qadence/test_error_models.py:135]: Remove

**Comment by Roland-djee** [comment]
> Thanks, @gvelikova and @Roland-djee. It looks quite good to me. A couple of general comments:
> 
> * I am not a big fan of the name `Errors`, I would much rather have the usual `Noise` which is much more common and understandable in the community.
 
Alright, let's go for `Noise` all the way through. Is it fine with you @gvelikova ?

> * I think the warning messages and compatibility of various options should be improved. For example, simulated noise now works for both `sample` and `expectation`, but only if measurement protocols are selected for the latter. This should be better communicated to the user and centralized in a single place, I think.

Sure, we can definitely do better docstrings.

> * Documentation is missing. Remember that one of the major points which decreased usage of shot-based measurements in the past was that there was practically no documentation. This really needs to be addressed.

Indeed, I started something in that direction but docs builds succeed locally but fail on CI. I'll investigate why.

> * I see why an `ExperimentComposer` will be needed if we add also, e.g., error mitigation to the various configurations. It can just be a single separate option to the backends/models called, e.g., `Protocols` or `RealisticModel` where we centralize the measurements, errors and mitigation.

And also for previous (`diff_mode` should go there I think) and forthcoming parameters (say, the `diff_engine` ?). Overall, we should discuss that but I think it goes under a pydantic model.

File: qadence/backend.py
Comment (Roland-djee) [reply qadence/backend.py:29]: Yes, I realised that after posting that comment.

File: qadence/backends/pyqtorch/backend.py
Comment (Roland-djee) [reply qadence/backends/pyqtorch/backend.py:190]: OK I can add a comment in the code.

File: qadence/backend.py
Comment (Roland-djee) [reply qadence/backend.py:23]: OK, good to me.

File: qadence/backends/braket/backend.py
Comment (Roland-djee) [reply qadence/backends/braket/backend.py:178]: Done.

File: qadence/errors/protocols.py
Comment (Roland-djee) [reply qadence/errors/protocols.py:12]: Will remove the comment unless you've got a better idea @madagra ?

File: qadence/errors/readout.py
Comment (Roland-djee) [reply qadence/errors/readout.py:87]: There is now an RFC for this @madagra.

File: qadence/backends/braket/backend.py
Comment (Roland-djee) [inline qadence/backends/braket/backend.py:181]:
```suggestion
                f"Error of type {error} are not implemented for expectation yet. "
```

File: qadence/execution.py
Comment (Roland-djee) [inline qadence/execution.py:140]: Why not using the pipe union ?

File: tests/qadence/test_error_models.py
Comment (Roland-djee) [reply tests/qadence/test_error_models.py:142]: Done.

File: qadence/backends/pyqtorch/backend.py
Comment (Roland-djee) [reply qadence/backends/pyqtorch/backend.py]: Done.

File: qadence/execution.py
Comment (gvelikova) [reply qadence/execution.py]: I was receiving complaints when running the CI, but maybe it was something else at the time?

File: qadence/backends/braket/backend.py
Comment (gvelikova) [reply qadence/backends/braket/backend.py]: I made it to be both Noise here, still, can turn it to "Error of type {noise} ..."

**Comment by gvelikova** [comment]
I guess the QNN would also need the noise added to .expectation and docstring?

File: docs/realistic_sims/noise.md
Comment (gvelikova) [inline docs/realistic_sims/noise.md:1]: Thanks @Roland-djee. Should we also add a .sample example?

Comment (Roland-djee) [reply docs/realistic_sims/noise.md:1]: Not sure we maintain/dev examples anymore. I think anything sits preferably in the docs now.

File: qadence/backend.py
Comment (gvelikova) [inline qadence/backend.py:171]: Why here the docstring was shortened?

File: qadence/execution.py
Comment (gvelikova) [inline qadence/execution.py:114]: The same as above.

File: qadence/backend.py
Comment (Roland-djee) [reply qadence/backend.py:171]: Well, it's been broken into few lines to comply with the docstring formatter. The first line has been shortened just to comply with the line length.

File: qadence/execution.py
Comment (Roland-djee) [reply qadence/execution.py:114]: Here too, compliant with the formatter. I am not sure the other lines about single dispatch are really informative as it is obvious from the code.

File: qadence/noise/readout.py
Comment (nmheim) [inline qadence/noise/readout.py:7]: why are we working with `np.array`s in here? does it make sense to use `torch.Tensor` here for to stick to just one array/tensor type in qadence?

File: tests/qadence/test_error_models.py
Comment (nmheim) [inline tests/qadence/test_error_models.py:61]: are these tests deterministic? should we set a seed/make the test flaky to make sure its not sometimes failing?

File: qadence/noise/protocols.py
Comment (nmheim) [inline qadence/noise/protocols.py:17]: should we turn `protocol` into a `StrEnum`?

File: qadence/noise/readout.py
Comment (nmheim) [inline qadence/noise/readout.py:106]: does the torch `noise_distribution` really return an `np.array`?

File: tests/qadence/test_error_models.py
Comment (gvelikova) [reply tests/qadence/test_error_models.py:61]: This particular test is deterministic in a way as the probability to invert the bit strings is 1.0, so we know what to expect.

File: qadence/noise/protocols.py
Comment (gvelikova) [reply qadence/noise/protocols.py:17]: Hm, I don't think we'd be changing this in this PR. Is this correct @Roland-djee?

Comment (Roland-djee) [reply qadence/noise/protocols.py:17]: Probably not in this one. But yes I think that would be useful. The only drawback is that users would need to import another type, here it is all encapsulated in `Noise`.

File: qadence/noise/readout.py
Comment (Roland-djee) [inline qadence/noise/readout.py:123]:
```suggestion
    from torch import Tensor
    err_idx: Tensor,
```

It is more idiomatic to Qadence style.

File: qadence/backends/braket/backend.py
Comment (madagra) [inline qadence/backends/braket/backend.py:177]: I would say: "for exact expectation". It is implemented in case of shot-based expectation.

File: qadence/backends/pulser/backend.py
Comment (madagra) [reply qadence/backends/pulser/backend.py:37]: Yes, exactly. You distinguish it from the file path. I would still use `__file__` here instead of `__name__`.

File: qadence/backends/pulser/backend.py
Comment (madagra) [inline qadence/backends/pulser/backend.py:282]: Same comment as above.

File: qadence/backends/pyqtorch/backend.py
Comment (madagra) [inline qadence/backends/pyqtorch/backend.py:192]: Again.

File: qadence/noise/readout.py
Comment (madagra) [inline qadence/noise/readout.py:145]: Name of the funciton is misleading, I think. I would call it `add_noise` for example (and also change the way it is called in the protocols enumeration then).

**Review by madagra — APPROVED** [review]
A few minor comments, but overall looks good to me @gvelikova and @Roland-djee.

File: qadence/backends/pulser/backend.py
Comment (Roland-djee) [reply qadence/backends/pulser/backend.py]: Done.

File: qadence/noise/readout.py
Comment (gvelikova) [inline qadence/noise/readout.py:53]: Here the output_type definition (same from the above function) is missing.

Comment (Roland-djee) [reply qadence/noise/readout.py:53]: I would suggest to avoid numpy types as much as you can. Sorry for not pointing that earlier.

Comment (gvelikova) [reply qadence/noise/readout.py:53]: I left these as an option, in case someone wants to use the bistring/tensor conversion as a standalone in the future. These are used with torch tensors internally.

Comment (gvelikova) [reply qadence/noise/readout.py:53]: OK then, I will remove these.

**Comment by Roland-djee** [comment]
Thanks @gvelikova. If it passes, you can hit the merge button :)

**Pull request merged**
